src/main/java/org/jenkinsci/plugins/docker/commons/fingerprint/DockerFingerprints.java

```diff
@@ -64,25 +64,54 @@ public static void addRunFacet(@Nonnull ContainerRecord record, @Nonnull Run<?,?
+     * Creates a new {@link DockerAncestorFingerprintFacet} and {@link DockerDescendantFingerprintFacet} and adds a run.
+     * Or adds to existing facets.
+     * @param ancestorImageId the ID of the image specified in a {@code FROM} instruction, or null in case of {@code scratch} (i.e., the descendant is a base image)
+     * @param descendantImageId the ID of the image which was built
+     * @param run the build in which the image building occurred
      */
-    public static void addFromFacet(@Nonnull String imageId, @Nonnull Run<?,?> run) throws IOException {
-        Fingerprint f = make(run, imageId);
+    public static void addFromFacet(@CheckForNull String ancestorImageId, @Nonnull String descendantImageId, @Nonnull Run<?,?> run) throws IOException {
+        long timestamp = System.currentTimeMillis();
+        if (ancestorImageId != null) {
+            Fingerprint f = make(run, ancestorImageId);
+            Collection<FingerprintFacet> facets = f.getFacets();
+            DockerAncestorFingerprintFacet ancestorFacet = null;
+            for (FingerprintFacet facet : facets) {
+                if (facet instanceof DockerAncestorFingerprintFacet) {
+                    ancestorFacet = (DockerAncestorFingerprintFacet) facet;
+                    break;
+                }
+            }
+            BulkChange bc = new BulkChange(f);
+            try {
+                if (ancestorFacet == null) {
+                    ancestorFacet = new DockerAncestorFingerprintFacet(f, timestamp, ancestorImageId);
+                    facets.add(ancestorFacet);
+                }
+                ancestorFacet.addDescendantImageId(descendantImageId);
+                ancestorFacet.addFor(run);
+                bc.commit();
+            } finally {
+                bc.abort();
+            }
+        }
+        Fingerprint f = make(run, descendantImageId);
         Collection<FingerprintFacet> facets = f.getFacets();
-        DockerFromFingerprintFacet fromFacet = null;
+        DockerDescendantFingerprintFacet descendantFacet = null;
         for (FingerprintFacet facet : facets) {
-            if (facet instanceof DockerFromFingerprintFacet) {
-                fromFacet = (DockerFromFingerprintFacet) facet;
+            if (facet instanceof DockerDescendantFingerprintFacet) {
+                descendantFacet = (DockerDescendantFingerprintFacet) facet;
                 break;
             }
         }
         BulkChange bc = new BulkChange(f);
         try {
-            if (fromFacet == null) {
-                fromFacet = new DockerFromFingerprintFacet(f, System.currentTimeMillis(), imageId);
-                facets.add(fromFacet);
+            if (descendantFacet == null) {
+                descendantFacet = new DockerDescendantFingerprintFacet(f, timestamp, descendantImageId);
+                facets.add(descendantFacet);
             }
-            fromFacet.addFor(run);
+            descendantFacet.addAncestorImageId(ancestorImageId);
```

Comment:
<code>ancestorImageId</code>  may be <code>null</code> here. Does it make sense to add nulls to the facet?